x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java

```diff
@@ -399,4 +488,16 @@ public FieldMapper.Builder getMergeBuilder() {
     protected String contentType() {
         return CONTENT_TYPE;
     }
+
+    @Override
+    protected SyntheticSourceSupport syntheticSourceSupport() {
+        var keepMode = sourceKeepMode();
+        if (keepMode.isPresent() && keepMode.get() != SourceKeepMode.NONE) {
```

Comment:
I think for now we need to remove `keepMode.isPresent() && ` from this if statement. 

The issue currently is that if `index.mapping.synthetic_source_keep` index setting is configured or `synthetic_source_keep` is specified at an object field at a higher level, then we don't know about this. The `sourceKeepMode()` only tells about the `synthetic_source_keep` of the current field.

I think we can fix this in a followup. I think we need to add logic to `MapperBuilderContext` to include what synthetic source keep is based on the index setting or any direct or indirect parent object fields. Let's explore this in a follow up PR, then we can get this PR in, knowing that the optimized synthetic source loader only is used if the `synthetic_source_keep` is set to `none` on a field of type `counted_keyword`.

Reply:
I agree that we need more logic to detect if synthetic_source_keep is set somewhere else and inherited here.

But if I just remove the `keepMode.isPresent() &&` from the if-statement, couldn't that potentially raise a `NoSuchElementException` if the keepMode is not specified in the field settings?

Sounds to me like we want the logic for now to only use optimized synthetic source if the `synthetic_source_keep` is explicitly `none`. I'll update the logic to instead be
```suggestion
        if (keepMode.isPresent() == false || keepMode.get() != SourceKeepMode.NONE) {
```

Reply:
Yes, my code suggestion is incorrect. We should fallback is keepMode isn't present or not set to none, like you suggest here.